lib/sites/cvs.rb

```diff
@@ -0,0 +1,202 @@
+require 'rest-client'
+
+require_relative './base_clinic'
+
+module Cvs
+  STATE = 'MA'
+  USER_AGENTS = []
+  CVS_CITIES = {}
+
+  class CvsCity
+    attr_reader :city, :stores, :should_tweet
+
+    def initialize(city, stores, should_tweet)
+      @city = city
+      @stores = stores
+      @should_tweet = should_tweet
+    end
+  end
+
+  File.open("user_agents.txt", "r") do |f|
+    f.each_line do |line|
+      USER_AGENTS.append(line.strip)
+    end
+  end
+
+  File.open("cvs_locations.csv", "r") do |f|
+    f.each_line do |line|
+      entry = line.strip.split(',')
+      city = entry[0]
+      stores = entry[1].to_i
+      should_tweet = entry[2] == "1"
+      CVS_CITIES[city] = CvsCity.new(city, stores, should_tweet)
+    end
+  end
+
+  # This array also defines the order in which city names appear, the order is determined
+  # by the number of stores in a city.
+  TWEET_ALLOWED_CITIES = CVS_CITIES.keys.filter { |cvs_city| CVS_CITIES[cvs_city].should_tweet }
+                                   .sort_by { |cvs_city| -1 * CVS_CITIES[cvs_city].stores }
+
+  def self.state_clinic_representation(storage, logger)
+    # For now, CVS is counted as one "clinic" for the whole state and every city offering
+    # with stores offering the vaccine is counted as one "appointment".
+    cvs_client = CvsClient.new(STATE, USER_AGENTS)
+    cvs_client.init_session(logger)
+
+    cities_with_appointments = cvs_client.cities_with_appointments(logger)
+    if cities_with_appointments.any?
+      logger.info "[CVS] There are #{cities_with_appointments.length} cities with appointments"
+      logger.info "[CVS] Cities with appointments: #{cities_with_appointments.join(", ")}"
+    else
+      logger.info "[CVS] No availability for any city in #{STATE}"
+    end
+
+    clinic = StateClinic.new(storage, cities_with_appointments, STATE)
+
+    return [clinic]
+  end
+
+  class StateClinic < BaseClinic
+    LAST_SEEN_CITIES_KEY = "last-cities".freeze
+
+    attr_accessor :appointments
+
+    def initialize(storage, cities, state)
+      super(storage)
+      @cities = cities.sort_by { |city| -1 * (CVS_CITIES[city].nil? ? 1 : CVS_CITIES[city].stores) }
+      @state = state
+      @appointments = @cities.length
+    end
+
+    def title
+      "CVS stores in #{@state}"
+    end
+
+    def appointments
+      @cities.length
+    end
+
+    def last_appointments
+      last_cities.length
+    end
+
+    def new_appointments
+      new_cities.length
+    end
+
+    def link
+      "https://www.cvs.com/immunizations/covid-19-vaccine"
+    end
+
+    def sign_up_page
+      link
+    end
+
+    def twitter_text
+      cities_text = ''
+      tweet_allowed_new_cities.each do |city|
+        if cities_text.length > 125
+          break
+        end
+        if cities_text.empty?
+          cities_text = city
+        else
+          cities_text = cities_text + ", " + city
+        end
+      end
+
+      # 30 chars + up to 125 chars + 85 chars = 240 chars, limit is 280
```

Comment:
Links are [shortened to 23 characters](https://developer.twitter.com/en/docs/counting-characters) automatically by Twitter's link-shortening service, so we should have a bit more to work with. I think we can also get this text to be a bit more informative with something like this where we say the number of other cities available:

```ruby
new_cities = tweet_allowed_new_cities
cities_text = new_cities.shift
while city = new_cities.shift do
  pending_text = ", #{city}"
  if cities_text.length + pending_text.length > 177
    cities_text += ", and #{new_cities.length + 1} others"
    break
  else
    cities_text += pending_text
  end
end
```